aiidalab_widgets_base/structures.py

```diff
@@ -180,33 +193,71 @@ def _default_node_class():
 
     @observe('node_class')
     def _change_structure_node(self, _=None):
-        self._structure_changed()
+        self.set_structure_node()
 
-    @validate('structure')
-    def _valid_structure(self, change):
-        """Returns ASE atoms object and sets structure_node trait."""
+    def set_structure_node(self):
+        """Set structure_node trait using the currently provided info."""
+        StructureNode = DataFactory(self.SUPPORTED_DATA_FORMATS[self.node_class])  # pylint: disable=invalid-name
+        structure_node = None
+        if len(self.history) > 1:
+            # There are some modifications, so converting from ASE.
+            structure_node = StructureNode(ase=self.structure)
+        else:
+            # If the input_structure trait is set to Atoms object, structure node must be created from it.
+            if isinstance(self.input_structure, Atoms):
+                structure_node = StructureNode(ase=self.input_structure)
+
+            # If the input_structure trait is set to AiiDA node, check what type
+            elif isinstance(self.input_structure, Data):
+
+                # Transform the structure to the StructureNode if needed.
+                if isinstance(self.input_structure, StructureNode):
+                    structure_node = self.input_structure
+                else:
+                    # self.structure was already converted to the ASE Atoms object.
+                    structure_node = StructureNode(ase=self.structure)
```

Comment:
I'm inclined to suggest that we refactor this converter code into a dedicated private class method: `_convert_to_structure_node(cls, structure)` and then call here with `structure_node = self._convert_to_structure_node(self.input_srtucture)`.